restic prune: selection of packs to repack based on size

wplapper opened the pull request:
cmd_prune: Selection of packs to repack based on size. Add option --small-pack-size so prune can repack all packfiles smaller than small-pack-size bytes.

<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.
-->

What does this PR change? What problem does it solve?
-----------------------------------------------------

<!--
Describe the changes and their purpose here, as detailed as needed.
-->
Add option --small-pack-size so prune can repack all packfiles smaller than small-pack-size bytes.
The new option is passed onto internal/repository/prune.go.
A new decision has been added to "decide what to do" to
move the intended packfiles onto the repackSmallCandidates queue.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------
https://github.com/restic/restic/issues/5109
closes https://github.com/restic/restic/issues/5109
<!--
Link issues and relevant forum posts here.

If this PR resolves an issue on GitHub, use "Closes #1234" so that the issue
is closed automatically when this PR is merged.
-->

Checklist
---------

<!--
You do not need to check all the boxes below all at once. Feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box. Enable a checkbox by replacing [ ] with [x].

Please always follow these steps:
- Read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- Enable [maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- Run `gofmt` on the code in all commits.
- Format all commit messages in the same style as [the other commits in the repository](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
-->

- [x] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I'm done! This pull request is ready for review.

MichaelEischer commented on cmd/restic/cmd_prune.go:
```diff
@@ -74,10 +77,11 @@ func init() {
 	f := c.Flags()
 	f.StringVar(&pruneOptions.MaxUnused, "max-unused", "5%", "tolerate given `limit` of unused data (absolute value in bytes with suffixes k/K, m/M, g/G, t/T, a value in % or the word 'unlimited')")
-	f.StringVar(&pruneOptions.MaxRepackSize, "max-repack-size", "", "stop after repacking this much data in total (allowed suffixes for `size`: k/K, m/M, g/G, t/T)")
+	f.StringVar(&pruneOptions.MaxRepackSize, "max-repack-size", "", "maximum `size` to repack (allowed suffixes: k/K, m/M, g/G, t/T)")
```
Please revert this change, the wording has somewhat recently changed to `stop after repacking this much data in total` to improve the clarity of the description.

wplapper replied:
done

MichaelEischer commented on cmd/restic/cmd_prune.go:
```diff
@@ -74,10 +77,11 @@ func init() {
 	f.BoolVar(&pruneOptions.RepackCacheableOnly, "repack-cacheable-only", false, "only repack packs which are cacheable")
 	f.BoolVar(&pruneOptions.RepackSmall, "repack-small", false, "repack pack files below 80% of target pack size")
 	f.BoolVar(&pruneOptions.RepackUncompressed, "repack-uncompressed", false, "repack all uncompressed data")
+	f.StringVar(&pruneOptions.SmallPackSize, "small-pack-size", "0", "pack `below-limit` packfiles smaller than (allowed suffixes: k/K, m/M, g/G, t/T)")
```
The default value should be an empty string.

wplapper replied:
done

MichaelEischer commented on cmd/restic/cmd_prune.go:
```diff
@@ -149,11 +160,7 @@ func runPrune(ctx context.Context, opts PruneOptions, gopts GlobalOptions, term
-	}
-
-	ctx, repo, unlock, err := openWithExclusiveLock(ctx, gopts, opts.DryRun && gopts.NoLock)
+	ctx, repo, unlock, err := openWithExclusiveLock(ctx, gopts, false)
```
Please revert this change.

wplapper replied:
done

MichaelEischer commented on internal/repository/prune.go:
```diff
@@ -352,7 +356,6 @@ func decidePackAction(ctx context.Context, opts PruneOptions, repo *Repository,
 				id.Str(), p.unusedSize+p.usedSize, packSize)
 			return ErrSizeNotMatching
 		}
-
```
unnecessary whitespace change.

wplapper replied:
done

MichaelEischer commented on internal/repository/prune.go:
```diff
@@ -385,12 +388,18 @@ func decidePackAction(ctx context.Context, opts PruneOptions, repo *Repository,
 			// if this is a data pack and --repack-cacheable-only is set => keep pack!
 			stats.Packs.Keep++
 
+		//wpl 2024-11-08 - repack when packSize is smaller than opts.SmallPackBytes
```
There is no need to manually mark who and when some logic was changed, git already takes care of that automatically.

wplapper replied:
comment removed.

MichaelEischer commented on internal/repository/prune.go:
```diff
@@ -401,7 +410,8 @@ func decidePackAction(ctx context.Context, opts PruneOptions, repo *Repository,
 		bar.Add(1)
 		return nil
-	})
+	}) // end repo.List(ctx, restic.PackFile ...)
```
Unnecessary comment.

wplapper replied:
comment removed

MichaelEischer commented on internal/repository/prune.go:
```diff
@@ -25,6 +25,9 @@ type PruneOptions struct {
 	MaxUnusedBytes func(used uint64) (unused uint64) // calculates the number of unused bytes after repacking, according to MaxUnused
 	MaxRepackBytes uint64
 
+	SmallPackSize  string
```
There's no point in passing the string version of SmallPackSize to `prune` as it only needs `SmallPackBytes`.

wplapper replied:
SmallPackSize removed

wplapper replied:
field SmallPackSize   removed

MichaelEischer commented on internal/repository/prune.go:
```diff
@@ -25,6 +25,9 @@ type PruneOptions struct {
 	MaxRepackBytes uint64
 
+	SmallPackSize  string
+	SmallPackBytes uint64
```
Please add a check to prune that `SmallPackBytes <= repo.packSize()`

wplapper replied:
added check in decidePackAction

MichaelEischer replied:
That check should happen right at the start of `PlanPrune`. `decidePackAction(...)` is too late as prune can have performed a significant amount of processing up to that point.

MichaelEischer commented on internal/repository/prune.go:
```diff
@@ -385,12 +388,18 @@ func decidePackAction(ctx context.Context, opts PruneOptions, repo *Repository,
 			stats.Packs.Keep++
 
+		//wpl 2024-11-08 - repack when packSize is smaller than opts.SmallPackBytes
+		case opts.RepackSmall && (uint64(packSize) <= opts.SmallPackBytes):
```
I don't think we need a special case here. Just set `targetPackSize` to `opts.SmallPackBytes` if the latter is non-zero and `opts.RepackSmall` is set.

MichaelEischer commented on internal/repository/prune.go:
```diff
@@ -434,12 +444,15 @@ func decidePackAction(ctx context.Context, opts PruneOptions, repo *Repository,
+	// calculate limit for number of unused bytes in the repo after repacking
+	maxUnusedSizeAfter := opts.MaxUnusedBytes(stats.Size.Used)
+	// wpl we need a mixed calculation based the existing condition and on --max-unused
+	if len(repackSmallCandidates) >= 10 || opts.RepackSmall && (uint64(sizeRepackSmallCandidates) >= maxUnusedSizeAfter) {
```
Please revert this change. If users set `--max-unused 0`, your change causes repeated prune executions to repack the same data each time. The point of repacking small files is reduce the number of files, but a few files more or less won't matter. If `--max-unused` yields a large value (>= 1.28GB to be precise), then the new clause is irrelevant as `len(repackSmallCandidates) >= 10` will always be true first.

MichaelEischer commented on changelog/unreleased/issue-5109:
```diff
@@ -0,0 +1,10 @@
+Issue #5109 - Selection of packs to repack based on size
```
Please follow the template from https://github.com/restic/restic/blob/master/changelog/TEMPLATE .
The changelog for this PR should look something like the following:
```
Enhancement: Make small pack size configureable for `prune`

The `prune` command now supports the `--small-pack-size` option that
allows repacking pack files smaller than the given size.

https://github.com/restic/restic/issues/5109
https://github.com/restic/restic/pull/5183
```

The changelog entries for restic are intended for end users. Those likely only know the CLI but have no knowledge of any internal code.

wplapper replied:
done.

MichaelEischer reviewed (changes requested):
The change in `internal/repository/prune.go` is more complex than it needs to be. It should be sufficient to just set the  `targetPackSize` to the user specified value. The current approach feels like a duplication of the existing selection of small packs.

The PR also must be rebased onto the current master branch as we generally don't merge PRs that contain merge commits. (See the `git rebase` command / the [git book](https://git-scm.com/book/en/v2/Git-Branching-Rebasing)).

I'd also like to have a small test for the new option, although I can also help with that. The test should probably use an approach similar to that of https://github.com/restic/restic/pull/5212/files .

MichaelEischer commented on internal/repository/prune.go:
```diff
@@ -374,6 +378,9 @@ func decidePackAction(ctx context.Context, opts PruneOptions, repo *Repository,
 		}
 
 		// decide what to do
+		if opts.SmallPackBytes > 0 && opts.RepackSmall {
```
Please move that to the point where `targetPackSize`  is initially computed (that is after line 336)

wplapper replied:
done

MichaelEischer commented on cmd/restic/cmd_prune.go:
```diff
@@ -136,6 +140,13 @@ func verifyPruneOptions(opts *PruneOptions) error {
 	}
 
+	smallPackSize := strings.TrimSpace(opts.SmallPackSize)
+	size, err := ui.ParseBytes(smallPackSize)
```
This needs a special case to not fail if  `smallPackSize` is an emtpy string.

wplapper replied:
fixed the bug.

MichaelEischer reviewed (changes requested):
Thanks, I found a few more nits

wplapper commented:
I added logic to the case `p.unusedBlobs == 0 && p.tpe != restic.InvalidBlob && !mustCompress`: in case we deal with `--repack-small` and have `--small-pack-size`with a non-zero value, these packs get all dropped into the repackSmallCandidates` bucket.

wplapper commented:
> I'd also like to have a small test for the new option, although I can also help with that. The test should probably use an approach similar to that of https://github.com/restic/restic/pull/5212/files .

The plan for the test is as follows:
1.) create repository  with packsize of 2M.
2.) create enough data for 3 packfiles, something in the range of 7-8 M.
3.) run a `repository.PlanPrune(...)` with a packsize of 16M (current default).
4.) run `plan.Execute(...)`, extract `pan.Stats()`.
5.) The result should be 2 packfiles, one for TreeBlobs and one for DataBlobs.
6.) Check that all blobs are contained in the new packfiles.

wplapper commented:
> 1.) create repository with packsize of 2M.
> 2.) create enough data for 3 packfiles, something in the range of 7-8 M.
> 3.) run a `repository.PlanPrune(...)` with a packsize of 16M (current default).
> 4.) run `plan.Execute(...)`, extract `plan.Stats()` and check.
> 5.) The result should be 2 packfiles, one for TreeBlobs and one for DataBlobs.
> 6.) Check that all blobs are contained in the new packfiles.

The reason why I change the packsize is the simulation to larger packsizes. There might be other methods of forcing repacking of smaller packfiles, but that's what I did :)

MichaelEischer commented on cmd/restic/cmd_prune.go:
```diff
@@ -78,6 +81,7 @@ func addPruneOptions(c *cobra.Command, pruneOptions *PruneOptions) {
 	f.BoolVar(&pruneOptions.RepackCacheableOnly, "repack-cacheable-only", false, "only repack packs which are cacheable")
 	f.BoolVar(&pruneOptions.RepackSmall, "repack-small", false, "repack pack files below 80% of target pack size")
 	f.BoolVar(&pruneOptions.RepackUncompressed, "repack-uncompressed", false, "repack all uncompressed data")
+	f.StringVar(&pruneOptions.SmallPackSize, "small-pack-size", "", "pack `below-limit` packfiles smaller than (allowed suffixes: k/K, m/M, g/G, t/T)")
```
I just noticed that the original issue suggested `--repack-smaller-than` as option. Is there a particular reason to deviate from that?

P.S. the description should be  (pack -> repack, packfiles -> pack files)
```
"repack pack files smaller than `below-limit` (allowed suffixes: k/K, m/M, g/G, t/T)"
```
We could also drop the `t/T` suffix as it will always lead to an error. `g/G` currently is also always an error as the pack size is still limited to 128MB, but that could change in the future.

MichaelEischer reviewed (changes requested):
I found a few issues with the test, see below. We also have to discuss the method name, in the linked issue the suggestion was to use `--repack-smaller-than`.

```
if packSize >= int64(targetPackSize) && opts.SmallPackBytes == 0 && !opts.RepackSmall {
```
As mentioned in the comment, I have no clue what this is supposed to do.

wplapper commented on internal/repository/prune.go:
```diff
@@ -326,7 +330,10 @@ func decidePackAction(ctx context.Context, opts PruneOptions, repo *Repository,
 	// only repack very small files by default
 	targetPackSize := repo.packSize() / 25
-	if opts.RepackSmall {
+	if opts.SmallPackBytes > 0 && opts.RepackSmall {
```
done.

wplapper commented on internal/repository/prune.go:
```diff
@@ -386,7 +393,7 @@ func decidePackAction(ctx context.Context, opts PruneOptions, repo *Repository,
 
 		case p.unusedBlobs == 0 && p.tpe != restic.InvalidBlob && !mustCompress:
-			if packSize >= int64(targetPackSize) {
+			if packSize >= int64(targetPackSize) && opts.SmallPackBytes == 0 && !opts.RepackSmall {
```
removed faultt if-clause.

wplapper commented on internal/repository/repository.go:
```diff
@@ -148,6 +148,11 @@ func (r *Repository) packSize() uint {
 }
 
+// SetPackSize - this is a hack!
+func (r *Repository) SetPackSize(size uint) {
```
method removed.

wplapper commented on internal/repository/prune_test.go:
```diff
@@ -191,3 +192,92 @@ func TestPruneMaxUnusedDuplicate(t *testing.T) {
+		return nil
+	}, &progress.NoopPrinter{})
+	rtest.OK(t, err)
+	_ = plan
```
assignment is gone.

wplapper commented on internal/repository/prune_test.go:
```diff
@@ -191,3 +192,92 @@ func TestPruneMaxUnusedDuplicate(t *testing.T) {
+	//t.Log(fmt.Sprintf("stats.Size  %+v", stats.Size))
+	//t.Log(fmt.Sprintf("stats.Blobs %+v", stats.Blobs))
+	//t.Log(fmt.Sprintf("stats.Packs %+v", stats.Packs))
+	rtest.Equals(t, stats.Size.Used <= stats.Size.Repack, true, "all sizes ...")
```
still working on the tests.

wplapper replied:
tests are improved, The checks should be more easily to understand.

MichaelEischer commented:
I've rebased the PR (and tried to fix the merge conflicts while rebasing) to remove the clutter from the "files changes" tab on Github.

wplapper commented:
Hi Michael,

Thank you for hard work on this PR. As I indicated earlier, I still don't
know much about git and github, and therefore make too many unforced
mistakes. I would appreciate a few hints, what to do, what to avoid.

Thanks again,
Winfried

On Mon, 3 Feb 2025 at 21:32, Michael Eischer ***@***.***>
wrote:

> I've rebased the PR (and tried to fix the merge conflicts while rebasing)
> to remove the clutter from the "files changes" tab on Github.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/restic/restic/pull/5183#issuecomment-2632122928>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEO7MGKEXVY6ILWUNHJDHDD2N7N77AVCNFSM6AAAAABTWK65N2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMZSGEZDEOJSHA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

MichaelEischer commented:
> Thank you for hard work on this PR. As I indicated earlier, I still don't know much about git and github, and therefore make too many unforced mistakes. I would appreciate a few hints, what to do, what to avoid.

No problem, it was basically "just" a matter of rebasing the PR. However, due to the merge conflicts that required a bit of experience with `git rebase`. The "files added" tab on Github doesn't seem to handle merge commits particularly well,  such that it also shows changes that already exist in the repository. The best way to avoid that is to not add merge commits in a PR in the first place, but instead to rebase it once there's a conflict with other changes in the repository. (having merge commits in a PR is allowed by Github, but frowned upon by many projects including restic)

`git rebase` allows modifying and cleaning up the commit history, which can be useful when the PR evolves over time. By removing unnecessary changes from the commit history, this also reduces the chance of having a merge commit that has to be addressed. Although that is probably already advanced git territory.

MichaelEischer commented:
Judging from the "rebase" commits in https://github.com/restic/restic/pull/5185 you're holding `git rebase` very wrong. I can recommend reading through the first few chapters of https://git-scm.com/book/en/v2 and trying out the examples along the way. `git rebase` is supposed to take the existing changes and apply them on top of a different (newer) state of the repository. Rebasing does not create additional commits  (I'm ignoring advanced git rebase usages here). Pushing those changes to Github also requires using `git push --force-with-lease`.

wplapper commented:
I understand now why my `git rebase` went so utterly wrong: I have a paper version of the ProGit book and it contains a critical omission in the `rebase` section and does not mention 
`$ git checkout master; $ git merge experiment` at all.

MichaelEischer commented:
Please stop pushing merge commits into your PRs.

wplapper commented:
I messed up git again, so I did a hard reset and activated all addition and changes.

wplapper commented:
Will be reworking the PR based on commit 8c12291f5

wplapper committed:
prune: make small pack size configureable for `prune` all changes together

cmd_prune.go: added option `--repack-smaller-than`
prune.go: added field `SmallPackBytes` to `PruneOptions`, including checking and processing
prune_test.go: added test `TestPruneSmall`
doc/060_forget.rst: added description of enhancement
changelog/unreleased/issue-5109: description of enhancement

wplapper committed:
restic prune changelog/unreleased - corrext typo

changed option from `--small-pack-size` to `--repack-smaller-than`

MichaelEischer committed:
prune: allow --repack-smaller-than independently of --repack-small

MichaelEischer committed:
prune: fix not working option

MichaelEischer reviewed (approved):
LGTM.

I've changed the `--repack-smaller-than` option to work independently of the `--repack-small` option. My goal is to eventually get rid of `--repack-small` so let's not tie those options together.

[Edit]See https://github.com/restic/restic/issues/5293 for more context on that[/Edit]
